File: ApiSuvesaPos/Negocio/Logica/ordencompra.vb
Namespace Logica
    Public Class ordencompra

        Private db As Datos.Class.OrdenCompra

        Sub New()
            Me.db = New Datos.Class.OrdenCompra
        End Sub

        Public Function Buscar(porNombre As Boolean, Filtro As String) As List(Of Datos.Models.Ordencompra)
            Dim datos As New List(Of Datos.Models.Ordencompra)
            datos = Me.db.Buscar(porNombre, Filtro)

            Dim Resultado As New List(Of Datos.Models.Ordencompra)
            For Each ord As Datos.Models.Ordencompra In datos
                Dim OrdenCompra As New Datos.Models.Ordencompra
                OrdenCompra.Orden = ord.Orden
                OrdenCompra.Proveedor = ord.Proveedor
                OrdenCompra.Nombre = ord.Nombre
                OrdenCompra.Fecha = ord.Fecha
                OrdenCompra.Contado = ord.Contado
                OrdenCompra.Credito = ord.Credito
                OrdenCompra.Diascredito = ord.Diascredito
                OrdenCompra.Plazo = ord.Plazo
                OrdenCompra.Descuento = ord.Descuento
                OrdenCompra.Impuesto = ord.Impuesto
                OrdenCompra.Total = ord.Total
                OrdenCompra.Observaciones = ord.Observaciones
                OrdenCompra.Usuario = ord.Usuario
                OrdenCompra.NombreUsuario = ord.NombreUsuario
                OrdenCompra.Entregar = ord.Entregar
                OrdenCompra.CodMoneda = ord.CodMoneda
                OrdenCompra.SubTotalGravado = ord.SubTotalGravado
                OrdenCompra.SubTotalExento = ord.SubTotalExento
                OrdenCompra.SubTotal = ord.SubTotal
                OrdenCompra.Anulado = ord.Anulado
                OrdenCompra.IdSucursal = ord.IdSucursal

                For Each det As Datos.Models.DetalleOrdencompra In Me.db.BuscarDetalle(OrdenCompra.Orden)
                    Dim Detalle As New Datos.Models.DetalleOrdencompra
                    Detalle.Id = det.Id
                    Detalle.Orden = det.Orden
                    Detalle.Codigo = det.Codigo
                    Detalle.Descripcion = det.Descripcion
                    Detalle.CostoUnitario = det.CostoUnitario
                    Detalle.Cantidad = det.Cantidad
                    Detalle.TotalCompra = det.TotalCompra
                    Detalle.PorcDescuento = det.PorcDescuento
                    Detalle.Descuento = det.Descuento
                    Detalle.PorcImpuesto = det.PorcImpuesto
                    Detalle.Impuesto = det.Impuesto
                    Detalle.OtrosCargos = det.OtrosCargos
                    Detalle.MontoFlete = det.MontoFlete
                    Detalle.Costo = det.Costo
                    Detalle.Gravado = det.Gravado
                    Detalle.Exento = det.Exento
                    Detalle.Vendidos = det.Vendidos
                    Detalle.ExistActual = det.ExistActual

                    OrdenCompra.DetalleOrdencompras.Add(Detalle)
                Next

                Resultado.Add(OrdenCompra)
            Next

            Return Resultado
        End Function

        Public Function Crear(ordencompra As Datos.Models.Ordencompra) As String
            Return Me.db.Crear(ordencompra)

            'Function RegistrarOrden() As Boolean

            '    If Me.SqlConnection1.State <> Me.SqlConnection1.State.Open Then Me.SqlConnection1.Open()
            '    Dim Trans As SqlTransaction = Me.SqlConnection1.BeginTransaction
            '    Try

            '        Me.Adapter_Orden_Compra.InsertCommand.Transaction = Trans
            '        Me.Adapter_Detalle_Orden_Compras.InsertCommand.Transaction = Trans

            '        Me.Adapter_Orden_Compra.UpdateCommand.Transaction = Trans
            '        Me.Adapter_Detalle_Orden_Compras.UpdateCommand.Transaction = Trans

            '        Me.Adapter_Orden_Compra.DeleteCommand.Transaction = Trans
            '        Me.Adapter_Detalle_Orden_Compras.DeleteCommand.Transaction = Trans

            '        Me.Adapter_Orden_Compra.Update(Me.DataSetOrden_Compras1, "ordencompra")
            '        Me.Adapter_Detalle_Orden_Compras.Update(Me.DataSetOrden_Compras1, "detalle_ordencompra")
            '        Trans.Commit()
            '        Return True

            '    Catch ex As Exception
            '        Trans.Rollback()
            '        MsgBox(ex.Message)
            '        Me.ToolBar1.Buttons(2).Enabled = True
            '        Return False
            '    End Try


            'End Function

        End Function

        Public Function Editar(id As Long, ordencompra As Datos.Models.Ordencompra) As String
            Dim res As String = Me.db.Editar(id, ordencompra)
            If res = "0" Then
                Return "No existe el valor"
            Else
                Return res
            End If
            'Function RegistrarOrden() As Boolean

            '    If Me.SqlConnection1.State <> Me.SqlConnection1.State.Open Then Me.SqlConnection1.Open()
            '    Dim Trans As SqlTransaction = Me.SqlConnection1.BeginTransaction
            '    Try

            '        Me.Adapter_Orden_Compra.InsertCommand.Transaction = Trans
            '        Me.Adapter_Detalle_Orden_Compras.InsertCommand.Transaction = Trans

            '        Me.Adapter_Orden_Compra.UpdateCommand.Transaction = Trans
            '        Me.Adapter_Detalle_Orden_Compras.UpdateCommand.Transaction = Trans

            '        Me.Adapter_Orden_Compra.DeleteCommand.Transaction = Trans
            '        Me.Adapter_Detalle_Orden_Compras.DeleteCommand.Transaction = Trans

            '        Me.Adapter_Orden_Compra.Update(Me.DataSetOrden_Compras1, "ordencompra")
            '        Me.Adapter_Detalle_Orden_Compras.Update(Me.DataSetOrden_Compras1, "detalle_ordencompra")
            '        Trans.Commit()
            '        Return True

            '    Catch ex As Exception
            '        Trans.Rollback()
            '        MsgBox(ex.Message)
            '        Me.ToolBar1.Buttons(2).Enabled = True
            '        Return False
            '    End Try


            'End Function

        End Function

        Public Function Anular(id As Long) As String
            Dim resp As String = Me.db.Anular(id)
            If resp = "0" Then
                Return "No existe el valor"
            Else
                Return resp
            End If
            'Function AnularOrden()
            '    Try
            '        Dim resp As Integer
            '        If Me.BindingContext(Me.DataSetOrden_Compras1, "ordencompra").Count > 0 Then
            '            If Me.BindingContext(Me.DataSetOrden_Compras1, "ordencompra.ordencompradetalle_ordencompra").Count > 0 Then

            '                resp = MessageBox.Show("¿Desea Anular esta Orden de Compra...?", "Seepos", MessageBoxButtons.YesNo, MessageBoxIcon.Question, MessageBoxDefaultButton.Button1)
            '                If resp = 6 Then
            '                    CheckBox1.Checked = True
            '                    Me.BindingContext(Me.DataSetOrden_Compras1, "ordencompra").EndCurrentEdit()

            '                    Me.DataSetOrden_Compras1.AcceptChanges()
            '                    MsgBox("La Orden de Compra ha sido anulada correctamente", MsgBoxStyle.Information)
            '                    Me.DataSetOrden_Compras1.detalle_ordencompra.Clear()
            '                    Me.DataSetOrden_Compras1.ordencompra.Clear()

            '                    Me.ToolBar1.Buttons(4).Enabled = True
            '                    Me.ToolBar1.Buttons(1).Enabled = True

            '                    Me.ToolBar1.Buttons(0).Text = "Nuevo"
            '                    Me.ToolBar1.Buttons(0).ImageIndex = 0
            '                    Me.ToolBar1.Buttons(3).Enabled = False
            '                    Me.ToolBar1.Buttons(2).Enabled = False
            '                    Me.ToolBar1.Buttons(4).Enabled = False


            '                    Me.txtUsuario.Enabled = True
            '                    Me.txtUsuario.Text = ""
            '                    Me.txtNombreUsuario.Text = ""
            '                    Me.txtUsuario.Focus()
            '                Else : Exit Function

            '                End If
            '            End If
            '        End If

            '    Catch ex As System.Exception
            '        MsgBox(ex.Message)
            '    End Try
            'End Function

        End Function

    End Class
End Namespace

File: ApiSuvesaPos/Datos/Class/OrdenCompra.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using Datos.Models;
using Microsoft.EntityFrameworkCore;

namespace Datos.Class
{
	public class OrdenCompra
	{
		private SeePOSContext entities;

		public OrdenCompra()
		{
			entities = new SeePOSContext();
		}

		public int Crear(Models.Ordencompra ordencompra) // registro de OrdenCompra
		{
			try
			{
				entities.Ordencompras.Add(ordencompra);
				return entities.SaveChanges();
			}
			catch (Exception ex)
			{
				throw ex;
			}
		}

		public List<Models.DetalleOrdencompra> BuscarDetalle(long Id)  //consultar OrdenCompra
		{
			try
			{
				List<Models.DetalleOrdencompra> result;
					var temp = from c in entities.DetalleOrdencompras
							   where c.Orden == Id
							   select c;
					result = temp.ToList<Models.DetalleOrdencompra>();
				
				if (result.Count > 0)
				{
					return result;
				}
				else
				{
					return result = null;
				}
			}
			catch (Exception ex)
			{
				throw ex;
			}
		}

		public List<Models.Ordencompra> Buscar(bool porNombre, string filtro)  //consultar OrdenCompra
		{
			try
			{
				List<Models.Ordencompra> result;
				if (porNombre == true)
				{
					var temp = from c in entities.Ordencompras
							   where c.Nombre.Contains(filtro)
							   orderby c.Fecha descending
							   select c;
					result = temp.ToList<Models.Ordencompra>();
				}
				else
				{
					var temp = from c in entities.Ordencompras
							   where c.Orden == Convert.ToInt32(filtro)
							   select c;
					result = temp.ToList<Models.Ordencompra>();
				}
				if (result.Count > 0)
				{
					return result;
				}
				else
				{
					return result = null;
				}
			}
			catch (Exception ex)
			{
				throw ex;
			}
		}

		public int Editar(long id, Models.Ordencompra Orden) //editar OrdenCompra
		{
			try
			{
				var p = entities.Ordencompras.Find(id);
				Models.Ordencompra OrdenCompra = p;
				if (OrdenCompra != null)
				{					
					OrdenCompra.Proveedor = Orden.Proveedor;
					OrdenCompra.Nombre = Orden.Nombre;
					OrdenCompra.Fecha = Orden.Fecha;
					OrdenCompra.Contado = Orden.Contado;
					OrdenCompra.Credito = Orden.Credito;
					OrdenCompra.Diascredito = Orden.Diascredito;
					OrdenCompra.Plazo = Orden.Plazo;
					OrdenCompra.Descuento = Orden.Descuento;
					OrdenCompra.Impuesto = Orden.Impuesto;
					OrdenCompra.Total = Orden.Total;
					OrdenCompra.Observaciones = Orden.Observaciones;
					OrdenCompra.Usuario = Orden.Usuario;
					OrdenCompra.NombreUsuario = Orden.NombreUsuario;
					OrdenCompra.Entregar = Orden.Entregar;
					OrdenCompra.CodMoneda = Orden.CodMoneda;
					OrdenCompra.SubTotalGravado = Orden.SubTotalGravado;
					OrdenCompra.SubTotalExento = Orden.SubTotalExento;
					OrdenCompra.SubTotal = Orden.SubTotal;
					OrdenCompra.Anulado = Orden.Anulado;

					var ac = from x in entities.DetalleOrdencompras
							 where x.Orden == id && !(from t in Orden.DetalleOrdencompras select t.Id).Contains(x.Id)
							 select x;
					List<Models.DetalleOrdencompra> Eliminar = ac.ToList<Models.DetalleOrdencompra>();

					foreach (Models.DetalleOrdencompra item in Eliminar)
					{
						var del = entities.DetalleOrdencompras.Find(item.Id);
						entities.Remove(del);
						entities.SaveChanges();
					}

					Models.DetalleOrdencompra nuevaLinea;
					foreach (Models.DetalleOrdencompra Detalle in OrdenCompra.DetalleOrdencompras)
					{
						//Agrega nuevos registros
						if (Detalle.Id == 0)
						{
							nuevaLinea = new Models.DetalleOrdencompra();
							//nuevaLinea.Id = Detalle.Id;
							nuevaLinea.Orden = Detalle.Orden;
							nuevaLinea.Codigo = Detalle.Codigo;
							nuevaLinea.Descripcion = Detalle.Descripcion;
							nuevaLinea.CostoUnitario = Detalle.CostoUnitario;
							nuevaLinea.Cantidad = Detalle.Cantidad;
							nuevaLinea.TotalCompra = Detalle.TotalCompra;
							nuevaLinea.PorcDescuento = Detalle.PorcDescuento;
							nuevaLinea.Descuento = Detalle.Descuento;
							nuevaLinea.PorcImpuesto = Detalle.PorcImpuesto;
							nuevaLinea.Impuesto = Detalle.Impuesto;
							nuevaLinea.OtrosCargos = Detalle.OtrosCargos;
							nuevaLinea.MontoFlete = Detalle.MontoFlete;
							nuevaLinea.Costo = Detalle.Costo;
							nuevaLinea.Gravado = Detalle.Gravado;
							nuevaLinea.Exento = Detalle.Exento;
							nuevaLinea.Vendidos = Detalle.Vendidos;
							nuevaLinea.ExistActual = Detalle.ExistActual;
							OrdenCompra.DetalleOrdencompras.Add(nuevaLinea);
						}
						else
						{
							//Actualiza los detalles
							var a = entities.DetalleOrdencompras.Find(Detalle.Id);
							Models.DetalleOrdencompra lineaModificada = a;
							if (lineaModificada != null)
							{
								lineaModificada.Id = Detalle.Id;
								lineaModificada.Orden = Detalle.Orden;
								lineaModificada.Codigo = Detalle.Codigo;
								lineaModificada.Descripcion = Detalle.Descripcion;
								lineaModificada.CostoUnitario = Detalle.CostoUnitario;
								lineaModificada.Cantidad = Detalle.Cantidad;
								lineaModificada.TotalCompra = Detalle.TotalCompra;
								lineaModificada.PorcDescuento = Detalle.PorcDescuento;
								lineaModificada.Descuento = Detalle.Descuento;
								lineaModificada.PorcImpuesto = Detalle.PorcImpuesto;
								lineaModificada.Impuesto = Detalle.Impuesto;
								lineaModificada.OtrosCargos = Detalle.OtrosCargos;
								lineaModificada.MontoFlete = Detalle.MontoFlete;
								lineaModificada.Costo = Detalle.Costo;
								lineaModificada.Gravado = Detalle.Gravado;
								lineaModificada.Exento = Detalle.Exento;
								lineaModificada.Vendidos = Detalle.Vendidos;
								lineaModificada.ExistActual = Detalle.ExistActual;

								entities.Entry(lineaModificada).State = EntityState.Modified;
								entities.SaveChanges();
							}
						}

					}

					entities.Entry(OrdenCompra).State = EntityState.Modified;
					return entities.SaveChanges();
				}
				else
				{
					return 0;// no se encotro el registro solicitado.
				}
			}
			catch (Exception ex)
			{
				throw ex;
			}
		}

		public int Anular(long id) //anular OrdenCompra
		{
			try
			{
				var p = entities.Ordencompras.Find(id);
				Models.Ordencompra viejo = p;
				if (viejo != null)
				{
					viejo.Anulado = true;
					entities.Entry(viejo).State = EntityState.Modified;
					return entities.SaveChanges();
				}
				else
				{
					return 0;// no se encotro el registro solicitado.
				}
			}
			catch (Exception ex)
			{
				throw ex;
			}
		}

	}
}


